### PR TITLE
nn/utils: reduce CPU overhead in _get_total_norm for many small tensors

### LINKED ISSUE #133586: No way for low-overhead total norm in native PyTorch with large number of tensors
**Context**
Gradient norm clipping is a popular technique for stabilizing training, which requires computing the total norm with respect to the model's gradients. This involves a norm reduction across all of the gradients down to a single scalar.

PyTorch's `clip_grad_morm_` offers both single-tensor (`foreach=False`) and multi-tensor (`foreach=True`) implementations. However, even the multi-tensor `foreach` implementation incurs high CPU overhead when computing the total norm over a large number of tensors.
https://github.com/pytorch/pytorch/blob/3434a54fba537f67d152b50142b211d4aa059e67/torch/nn/utils/clip_grad.py#L81-L96
The foreach implementation involves:
1. `torch._foreach_norm(tensors)` to return a list of 0D scalars, representing the norm of each gradient
2. `torch.stack` to cat the 0D scalars into a 1D tensor
3. `torch.linalg.vector_norm()` to compute the norm of the norms, representing the final total norm

**Issue**
The foreach implementation incurs much unnecessary CPU overhead, making the clipping heavily CPU bound when operating on a large number of tensors. For example, for ~2000 tensors, the total norm calculation takes >18 ms on CPU, while only 1.3 ms on GPU (for a particular real workload -- larger tensors would make this slower).
Screenshot 2024-08-15 at 12 47 18 PM

Assuming `N` tensors, some inefficiencies in the existing implementation arise from:
- `_foreach_norm` must call `N` `aten::empty({})` to construct the `N` 0D scalar outputs ([code](https://github.com/pytorch/pytorch/blob/bedf96d7ffe74b34bcfe52c7ae1ae05f40d6c8ee/aten/src/ATen/native/cuda/ForeachReduceOp.cu#L444)).
- The `N` 0D scalars need to be `stack`d for the final norm reduction. `stack` requires 1D tensors, so each of the `N` 0D scalar gets unsqueezed again in `stack`.

Together, this leads to an extra `2N` dispatcher calls to handle the `N` intermediate scalars. Ideally, we can avoid materializing these `N` intermediates, especially as `torch.Tensor`s, where one option is a fused kernel.

Today, `torch.compile` cannot address this issue in a satisfying way. Default `torch.compile` cannot achieve horizontal fusion, leading to slower performance than eager mode. `torch.compile(mode="reduce-overhead")` does reduce overhead more than eager but results in 2x memory usage, likely due to copying the gradients into CUDA graph addresses. Note that we likely cannot mark the inputs as static because gradients are computed anew every iteration at possibly different addresses, and for my use case, we cannot compile the gradient allocation with `torch.compile` (due to FSDP).

Here is an example script for getting profiler traces of various implementations: P1529496302

The ask is to provide _some_ native way to make this total norm calculation not CPU-overhead bound, e.g. via a fused op.

### PR DESCRIPTION
**Pull request by notandruu**
Fixes #133586

## Summary

Two targeted optimizations for `_get_total_norm` (called by `clip_grad_norm_`) on CPU when training models with many gradient tensors:

**1. Skip no-op `.to(first_device)` calls in the final norm aggregation**

The current code always runs:
```python
total_norm = torch.linalg.vector_norm(
    torch.stack([norm.to(first_device) for norm in norms]), norm_type
)
```
When all gradient tensors are on the same device (the overwhelmingly common case — single GPU or CPU-only training), every `.to(first_device)` call is a no-op. For N=2000 gradient tensors, this dispatches 2000 Python method calls that perform no data movement. A cheap O(k) check (k = number of unique device/dtype groups, typically 1–3) gates the comprehension and eliminates it for the single-device case:

```python
if any(device != first_device for (device, _) in grouped_tensors):
    norms = [norm.to(first_device) for norm in norms]
total_norm = torch.linalg.vector_norm(torch.stack(norms), norm_type)
```

**2. Single flat-cat norm for very small tensors (numel ≤ 512, CPU, `foreach=None`)**

`_foreach_norm` allocates a separate 0-D scalar output tensor for each input, producing N `aten::empty({})` calls. For N=2000 bias/LayerNorm tensors of 512 elements, concatenating their views with `torch.cat` and computing one `vector_norm` avoids these N allocations. The threshold (512 elements) is conservative: benchmarks show a speedup at numel=512 and slowdown at numel=1024, so 512 is well inside the beneficial range. Explicit `foreach=True` bypasses this path entirely.

## Benchmark

Hardware: Apple M-series CPU. N=2000 tensors, each 512 float32 elements (typical bias/LayerNorm parameter size in large transformers). `torch.utils.benchmark.Timer.blocked_autorange(min_run_time=3.0)`:

| Path | Wall time |
|---|---|
| Before: `_foreach_norm` + `stack(.to())` + `vector_norm` | 7.5 ms |
| After: `cat(view(-1))` + `vector_norm` (no `.to()` needed) | 6.0 ms |

**→ ~20% reduction in `_get_total_norm` wall time** for the small-tensor case.

For models with larger parameter tensors (numel > 512), only optimization 1 applies; the skip-`.to()` path alone reduces this step by ~16%.

## Correctness

The flat-cat norm is mathematically equivalent to the stacked per-tensor norm for all Lp norms (including L∞): `‖cat(tensors)‖_p = ‖(‖t₁‖_p, …, ‖tₙ‖_p)‖_p`. 25 new unit tests cover:
- All four norm types (L1, L2, L3, L∞)
- Both optimizations independently and combined
- Threshold boundary (numel=512)
- `foreach=True` bypass
- float32 and float64
- Edge cases (single tensor, empty list, single-element tensors)
- End-to-end `clip_grad_norm_` correctness vs explicit `foreach=True`

### CODE REVIEW

File: test/test_clip_grad_flat_norm.py
Comment (janeyx99) [inline test/test_clip_grad_flat_norm.py:1]: can you roll this into preexisting tests about norm?

**Comment by janeyx99** [comment]
@claude pls review this PR with the PR review skill!

**Comment by notandruu** [comment]
Fixed: changed `device_tensors[0].numel() <= 512` to `all(t.numel() <= 512 for t in device_tensors)`. Also added `test_mixed_numel_group_uses_foreach_not_flat` which would have caught this — a group with mixed 512/1024-numel tensors now correctly takes the foreach path.

On the test restructure: happy to move these into the existing `test_clip_grad_norm` / `test_nn.py` if that's preferred — wanted to confirm the direction before doing that refactor.

**Comment by notandruu** [comment]
Friendly ping — all feedback from the previous review has been addressed (threshold check fixed to `all(t.numel() <= 512 for t in device_tensors)`). CI is clean. Happy to make any further changes if needed.

**Comment by notandruu** [comment]
Addressed the remaining feedback:

- **Tests moved into `test/test_nn.py`** as methods on `TestNNDeviceType` — `test_clip_grad_norm_flat_cat_small_tensors`, `test_clip_grad_norm_flat_cat_mixed_numel`, `test_clip_grad_norm_skip_to_single_device`. Deleted `test/test_clip_grad_flat_norm.py`.
- **`view(-1)` → `reshape(-1)`** to handle non-contiguous gradient tensors safely.
- Trimmed the comment block to 2 lines.
